electrum/lnrouter.py

```diff
@@ -161,12 +167,251 @@ def is_fee_sane(fee_msat: int, *, payment_amount_msat: int) -> bool:
+        # we don't want to record less significant info
+        # (not sendable amount is higher than known not sendable amount):
+        if self._cannot_send_forward and self._cannot_send_forward < amount:
+            return
+        self._cannot_send_forward = amount
+        if self._can_send_forward and self._can_send_forward > self._cannot_send_forward:
+            self._can_send_forward = None
+        # if we can't send over the channel, we should be able to send in the
+        # reverse direction
+        self.can_send_backward = amount
+
+    @property
+    def cannot_send_backward(self):
+        return self._cannot_send_backward
+
+    @cannot_send_backward.setter
+    def cannot_send_backward(self, amount):
+        if self._cannot_send_backward and self._cannot_send_backward < amount:
+            return
+        self._cannot_send_backward = amount
+        if self._can_send_backward and self._can_send_backward > self._cannot_send_backward:
+            self._can_send_backward = None
+        self.can_send_forward = amount
+
+    def can_send(self, is_forward_direction: bool):
+        # make info invalid after some time?
+        if is_forward_direction:
+            return self.can_send_forward
+        else:
+            return self.can_send_backward
+
+    def cannot_send(self, is_forward_direction: bool):
+        # make info invalid after some time?
+        if is_forward_direction:
+            return self.cannot_send_forward
+        else:
+            return self.cannot_send_backward
+
+    def update_can_send(self, is_forward_direction: bool, amount: int):
+        if is_forward_direction:
+            self.can_send_forward = amount
+        else:
+            self.can_send_backward = amount
+
+    def update_cannot_send(self, is_forward_direction: bool, amount: int):
+        if is_forward_direction:
+            self.cannot_send_forward = amount
+        else:
+            self.cannot_send_backward = amount
+
+    def __repr__(self):
+        return f"forward: can send: {self._can_send_forward}, cannot send: {self._cannot_send_forward}, \n" \
+               f"backward: can send: {self._can_send_backward} cannot send: {self._cannot_send_backward}, \n" \
+               f"blacklisted: {self.blacklisted}"
+
+
+class LiquidityHints:
+    """Implements liquidity hints for channels in the graph.
+
+    This class can be used to update liquidity information about channels in the
+    graph. Implements a penalty function for edge weighting in the pathfinding
+    algorithm that favors channels which can route payments and penalizes
+    channels that cannot.
+    """
+    # TODO: incorporate in-flight htlcs
+    # TODO: use timestamps for can/not_send to make them None after some time?
+    # TODO: hints based on node pairs only (shadow channels, non-strict forwarding)?
+    def __init__(self):
+        self.lock = RLock()
+        self._liquidity_hints: Dict[Tuple[bytes, bytes, ShortChannelID], LiquidityHint] = {}
```

Comment:
Why does the key contain both node ids and scid?
I first thought you only want the two nodes to determine direction but then I saw the nodes are ordered.

Reply:
Yes, I had it that way in the beginning and I'm still not 100% sure whether we should do hints per channel or per node pairs. I changed it now to key to the channel id only.